### PR TITLE
Clean up line tracking logic for debug info

### PR DESCRIPTION
**Pull request by qingyuanzNV**
This patch cleans up the line tracking logic when debug info is turned on, including:
1. Add correct file tracking. We used to have `sourceFileStringId` and `currentFileId` in the `SpvBuilder`, but both were always OpString id of the main file. Renamed them to `mainFileId` and `currentFileId` and implemented file tracking accordingly.
2. Make insertion of OpLine/OpDebugLine/OpDebugScope lazy. We used to insert OpLine right after setting the current debug source location, sometimes resulting in multiple cascading OpLines without any meaningful instructions in between. By making the process lazy, an OpLine is only inserted when an instruction is added with a dirty line tracker.
3. Added per-block line/scope tracker. So whenever an OpLine/OpDebugLine/OpScope is inserted, the state is stored in the `Block` instead of `SpvBuilder`. This is how the standard defines the lifetime of such instruction.

Also, some minor changes are:
1. Skip setting debug source location for linking object at the end of the AST
2. Rename some functions to better reflect that they are debug info related
3. No longer generate an OpLine preceding an OpFunction when Non-Semantic Debug is requested. The line information is available via the extension already.

~~As a result of this patch, we should have non-semantic debug reporting the correct file for variables, functions and .etc.~~
Hopefully, after this patch, we could have more accurate information to fix the file reporting for variables, functions and .etc for non-semantic debug extension. Also, this helps avoid a lot of cascading OpLines for a potential fix for #3403

### CODE REVIEW

**Comment by qingyuanzNV** [comment]
@arcady-lunarg We are trying to catch an internal deadline and put this patch in. Could you help review this PR? Thanks!

**Comment by arcady-lunarg** [comment]
Yes, I am going to review it today, it also fixes some issues for us. I was
at Vulkanised the past few days and didn't get a chance to look at it in
enough detail.

File: SPIRV/SpvBuilder.h [hunk 1]
```diff
@@ -409,11 +417,17 @@ class Builder {
     // Also reset current last DebugScope and current source line to unknown
     void setBuildPoint(Block* bp) {
         buildPoint = bp;
-        lastDebugScopeId = NoResult;
-        currentLine = 0;
+        // TODO: Technically, change of build point should set line tracker dirty. But we'll have bad line info for
+        //       branch instructions. Commenting this for now because at least this matches the old behavior.
+        // dirtyLineTracker = true;
```
Comment (arcady-lunarg) [inline SPIRV/SpvBuilder.h:422]: Do you have a pending change to clean this up? I don't really like the idea of checking in TODOs or commented code. Actually, you should just delete the commented out line, but keep the TODO comment for now.

**Review by arcady-lunarg — CHANGES_REQUESTED** [review]
Just a couple of nits, but overall I think this looks good, and the results seem fine. Also your PR description is great, and you should use some version of it as a commit message (and if not, I'll probably do "squash and merge" and do it anyway).

File: SPIRV/spvIR.h [hunk 2]
```diff
@@ -292,6 +320,12 @@ class Block {
     std::vector<std::unique_ptr<Instruction> > localVariables;
     Function& parent;
 
+    // Track source location of the last source location marker instruction.
+    std::optional<DebugSourceLocation> currentSourceLoc;
+
+    // Track scope of the last debug scope instruction.
+    std::optional<spv::Id> currentDebugScope;
```
[the remark this answers is not in the document]
Comment (qingyuanzNV) [reply SPIRV/spvIR.h:327]: Added assert.

Technically, there's an OpDebugNoScope to clear the current scope. But as we never generate the op and I don't see a use case of the op in glslang, I didn't add a way to clear the tracker to keep it simple.

File: SPIRV/SpvBuilder.h (hunk 1 above)
Comment (qingyuanzNV) [reply SPIRV/SpvBuilder.h]: Done

**Comment by qingyuanzNV** [comment]
Could you help squash and merge with the commit message below (so it's easier for you to review the new changes):
```
Clean up the debug line info tracking and generation. 
- Correctly populate the field `currentFileId` with the presence of include directive
- Support lazy OpLine/OpDebugLine generation only when a real instruction is added instead of a debug location is set
- Improve the debug location tracking to per-block instead of just per-builder
- A few bug fixes related to debug source info
```

**Comment by qingyuanzNV** [comment]
By the way, one important change of this PR is that I added an `addInstruction(...)` to `SpvBuilder` to hijack all instructions added to the current build point. You guys might need be alert on this when reviewing new changes to make sure that all new instructions added to the current build point should go through that function instead of directly calling `buildPoint.addInstruction(...)`.

**Review by arcady-lunarg — APPROVED** [review]
LGTM, thanks, this is a huge improvement.

**Pull request merged**